File: ObjCExampleApp/ViewController.m
//
//  ViewController.m
//  AAObjCExampleApp
//
//  Created by Brett Clifton on 8/10/20.
//  Copyright © 2020 AdAdapted. All rights reserved.
//

#import "ViewController.h"
#import "ObjCExampleApp-Swift.h"
@import AASwiftSDK;

@interface ViewController ()

@property (weak, nonatomic) IBOutlet AdAdaptedZoneView *adZone;
@property (weak, nonatomic) IBOutlet ObjCSearchTextField *searchTextField;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    defaultItems = [[NSMutableArray alloc]initWithObjects:@"Milk",@"Bread",@"Coffee",@"Eggs",@"Cheese", nil];
    listData = [[NSMutableArray alloc]initWithObjects:@"Milk",@"Bread", nil];
    listTableView.dataSource = self;
    listTableView.delegate = self;
    _searchTextField.delegate = self;
    
    [_adZone setZoneOwner:self];
    [AASDK registerContentListenersFor:self];
    
    _searchTextField.minCharactersNumberToStartFiltering = 3;
    [_searchTextField filterStrings: defaultItems];
    
}

- (void)aaContentNotification: (NSNotification*) notification {
    NSLog(@"In-app content available");
    NSString* adType = [[notification userInfo] objectForKey: AASDK.KEY_TYPE];
    NSString* zoneId = [[notification userInfo] objectForKey: AASDK.KEY_ZONE_ID];
    AdContent *adContent = [[notification userInfo] objectForKey: AASDK.KEY_AD_CONTENT];

    NSLog(@"AdType: %@", adType);
    NSLog(@"AdZone: %@", zoneId);
    for (AADetailedListItem *item in adContent.detailedListItems) {
        NSLog(@"Item Name: %@", item.productTitle);
        NSLog(@"Item Brand: %@", item.productBrand);
        NSLog(@"Item UPC: %@", item.productUpc);
        NSLog(@"Item Sku: %@", item.retailerSku);
        NSLog(@"Item Retailer ID: %@", item.retailerId);
        NSLog(@"Item Desc: %@", item.productDescription);
        [self insertItemToList: item.productTitle];
    }

    // Acknowledge the items were added to the list
    [adContent acknowledge];
}

- (UIViewController *)viewControllerForPresentingModalView {
    return self;
}


- (void)zoneViewDidLoadZone:(AAZoneView *)view {
    NSLog(@"Zone LOADED");
}

- (void)zoneViewDidFailToLoadZone:(AAZoneView *)view {
    NSLog(@"Zone FAILED TO LOAD");
}

- (void)aaPayloadNotification:(NSNotification*)notification {
    NSLog(@"Out-of-app content available");
    NSArray *adPayload = [notification.userInfo objectForKey:AASDK.KEY_CONTENT_PAYLOADS];

    for (AAContentPayload* payload in adPayload) {
        for (AADetailedListItem *item in payload.detailedListItems) {
            NSLog(@"********* AADetailedListItem: %@", item.productTitle);
            NSLog(@"%@", item.productCategory);
            NSLog(@"%@", item.productUpc);
            NSLog(@"%@", item.retailerId);
            NSLog(@"%@ ********", item.productDescription);
            [self insertItemToList:item.productTitle];
        }

        [payload acknowledge];
    }
}

//Non AASDK calls

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
     static NSString *cellId = @"SimpleTableId";
       
       UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellId];
       if (cell == nil) {
           cell = [[UITableViewCell alloc]initWithStyle:
                   UITableViewCellStyleDefault reuseIdentifier:cellId];
       }
    NSString *stringForCell;
    stringForCell= [listData objectAtIndex:indexPath.row];
    [cell.textLabel setText:stringForCell];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [listData count];
}

- (IBAction)addToList:(id)sender {
    if (_searchTextField.text.length != 0) {
        [self insertItemToList:_searchTextField.text];
        _searchTextField.text = @"";
    }
}

- (void)insertItemToList:(NSString*)itemName {
    [listData addObject:itemName];
    [listTableView reloadData];
}

- (BOOL)textFieldShouldReturn:(ObjCSearchTextField *)textField {
    [self addToList:_searchTextField];
    return YES;
}

@end

File: SwiftExampleApp/ViewController.swift
//
//  ViewController.swift
//  SwiftExampleApp
//
//  Created by Brett Clifton on 8/10/20.
//  Copyright © 2020 AdAdapted. All rights reserved.
//

import UIKit
import AASwiftSDK

class ViewController:
    UIViewController,
    UITableViewDataSource,
    UITableViewDelegate,
    UISearchTextFieldDelegate,
    AAZoneViewOwner,
    AASDKContentDelegate
{
    
    @IBOutlet weak var adAdaptedZoneView: AdAdaptedZoneView!
    @IBOutlet weak var addButton: UIButton!
    @IBOutlet weak var listTableView: UITableView!
    @IBOutlet weak var searchTextField: SearchTextField!
    
    var listData = [String]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        searchTextField.delegate = self
        listTableView.delegate = self
        listTableView.dataSource = self
        adAdaptedZoneView.setZoneOwner(self)
        AASDK.registerContentListeners(for: self)
        
        listData = ["Eggs", "Bread"]
        searchTextField.theme.font = UIFont.systemFont(ofSize: 15)
        searchTextField.minCharactersNumberToStartFiltering = 3
        searchTextField.filterStrings(getListItems())
    }
    
    //MARK: AA SDK Calls
    
    func viewControllerForPresentingModalView() -> UIViewController? {
        return self
    }
    
    func aaContentNotification(_ notification: Notification) {
        print("In-app content available")
        
        guard let userinfo = notification.userInfo else {
            print("userinfo is nil")
            return
        }
        
        guard let adContent = userinfo[AASDK.KEY_AD_CONTENT] as? AdContent else {
            print("userinfo[AASDK.KEY_AD_CONTENT] is nil")
            return
        }

        for item in adContent.detailedListItems {
            print("AADetailedListItem: \(item.productTitle), \(String(describing: item.productBrand)), \(String(describing: item.productUpc)), \(String(describing: item.retailerId)), \(String(describing: item.productCategory)), \(String(describing: item.productDescription))")
            appendListItem(itemName: item.productTitle)
        }

        // Acknowledge the items were added to the list
        adContent.acknowledge()
    }

    func zoneViewDidLoadZone(_ view: AAZoneView?) {
        if let zoneId = view?.zoneId {
            print("Zone " + zoneId + " loaded")
        }
    }

    func zoneViewDidFail(toLoadZone view: AAZoneView?) {
        print("Zone failed to load")
    }
    
    //Out of App AddIt
    func aaPayloadNotification(_ notification: Notification) {
        print("Out-of-app content available")
        guard let userinfo = notification.userInfo else { return }
        guard let adPayload = userinfo[AASDK.KEY_CONTENT_PAYLOADS] as? [AAContentPayload] else { return }

        for payload in adPayload {
            for item in payload.detailedListItems {
                print("AADetailedListItem: ", item.productTitle)
                appendListItem(itemName: item.productTitle)
            }

            payload.acknowledge()
        }
    }
    
    //MARK: Other Calls
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        searchTextField.resignFirstResponder()
        return true
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        addItemToList(addButton)
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return listData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell()
        let item = listData[indexPath.row]
        cell.textLabel?.text = item
        
        return cell
    }

    @IBAction func textFieldChanged(_ sender: UITextField) {
        searchTextField.filterStrings(getListItems())
    }
    
    @IBAction func addItemToList(_ sender: UIButton) {
        //test call to get recipe ads
        adAdaptedZoneView.setAdZoneContext(contextID: "RecipeID")
        //
        if searchTextField.text != nil && !searchTextField.text!.isEmpty {
            appendListItem(itemName: searchTextField.text!)
        }
    }
    
    private func getListItems(suggestion: String = String()) -> [String] {
        guard let path = Bundle.main.url(forResource: "DefaultListItems", withExtension: "plist") else {return []}
        let data = try! Data(contentsOf: path)
        
        guard var plist = try! PropertyListSerialization.propertyList(from: data, options: .mutableContainers, format: nil) as? [String] else {return []}
        
        if !suggestion.isEmpty {
            plist.append(suggestion)
        }
        
        return plist
    }
    
    private func appendListItem(itemName: String = "") {
        listData.append(itemName)
        listTableView.reloadData()
        searchTextField.text = ""
    }
}
